File: src/main/java/com/dh/ondot/schedule/api/PlaceController.java
package com.dh.ondot.schedule.api;

import com.dh.ondot.schedule.api.request.PlaceHistoryDeleteRequest;
import com.dh.ondot.schedule.api.request.PlaceHistorySaveRequest;
import com.dh.ondot.schedule.api.response.PlaceHistoryResponse;
import com.dh.ondot.schedule.api.response.PlaceSearchResponse;
import com.dh.ondot.schedule.api.swagger.PlaceSwagger;
import com.dh.ondot.schedule.application.PlaceFacade;
import com.dh.ondot.schedule.application.dto.PlaceSearchResult;
import jakarta.validation.Valid;
import jakarta.validation.constraints.NotBlank;
import lombok.RequiredArgsConstructor;
import org.springframework.http.HttpStatus;
import org.springframework.validation.annotation.Validated;
import org.springframework.web.bind.annotation.*;

import java.util.List;

@Validated
@RestController
@RequiredArgsConstructor
@RequestMapping("/places")
public class PlaceController implements PlaceSwagger {
    private final PlaceFacade placeFacade;

    @ResponseStatus(HttpStatus.OK)
    @GetMapping("/search")
    public List<PlaceSearchResponse> searchPlaces(
            @NotBlank(message = "query 파라미터는 필수입니다.")
            @RequestParam("query") String query
    ) {
        List<PlaceSearchResult> searchPlaces = placeFacade.searchPlaces(query);

        return PlaceSearchResponse.fromList(searchPlaces);
    }

    @ResponseStatus(HttpStatus.CREATED)
    @PostMapping("/history")
    public void saveHistory(
            @RequestAttribute("memberId") Long memberId,
            @Valid @RequestBody PlaceHistorySaveRequest request
    ) {
        placeFacade.saveHistory(memberId, request.title(),
                request.roadAddress(), request.longitude(), request.latitude());
    }

    @ResponseStatus(HttpStatus.OK)
    @GetMapping("/history")
    public List<PlaceHistoryResponse> history(
            @RequestAttribute("memberId") Long memberId
    ) {
        return PlaceHistoryResponse.fromList(placeFacade.getHistory(memberId));
    }

    @ResponseStatus(HttpStatus.NO_CONTENT)
    @DeleteMapping("/history")
    public void deleteHistory(
            @RequestAttribute("memberId") Long memberId,
            @Valid @RequestBody PlaceHistoryDeleteRequest request
    ) {
        placeFacade.deleteHistory(memberId, request.searchedAt());
    }
}

File: src/main/kotlin/com/dh/ondot/schedule/presentation/PlaceController.kt
package com.dh.ondot.schedule.presentation

import com.dh.ondot.schedule.presentation.request.PlaceHistoryDeleteRequest
import com.dh.ondot.schedule.presentation.request.PlaceHistorySaveRequest
import com.dh.ondot.schedule.presentation.response.PlaceHistoryResponse
import com.dh.ondot.schedule.presentation.response.PlaceSearchResponse
import com.dh.ondot.schedule.presentation.swagger.PlaceSwagger
import com.dh.ondot.schedule.application.PlaceFacade
import jakarta.validation.Valid
import jakarta.validation.constraints.NotBlank
import org.springframework.http.HttpStatus
import org.springframework.validation.annotation.Validated
import org.springframework.web.bind.annotation.DeleteMapping
import org.springframework.web.bind.annotation.GetMapping
import org.springframework.web.bind.annotation.PostMapping
import org.springframework.web.bind.annotation.RequestAttribute
import org.springframework.web.bind.annotation.RequestBody
import org.springframework.web.bind.annotation.RequestMapping
import org.springframework.web.bind.annotation.RequestParam
import org.springframework.web.bind.annotation.ResponseStatus
import org.springframework.web.bind.annotation.RestController

@Validated
@RestController
@RequestMapping("/places")
class PlaceController(
    private val placeFacade: PlaceFacade,
) : PlaceSwagger {

    @ResponseStatus(HttpStatus.OK)
    @GetMapping("/search")
    override fun searchPlaces(
        @NotBlank(message = "query 파라미터는 필수입니다.")
        @RequestParam("query") query: String,
    ): List<PlaceSearchResponse> {
        val searchPlaces = placeFacade.searchPlaces(query)

        return PlaceSearchResponse.fromList(searchPlaces)
    }

    @ResponseStatus(HttpStatus.CREATED)
    @PostMapping("/history")
    override fun saveHistory(
        @RequestAttribute("memberId") memberId: Long,
        @Valid @RequestBody request: PlaceHistorySaveRequest,
    ) {
        placeFacade.saveHistory(
            memberId, request.title ?: request.roadAddress,
            request.roadAddress, request.longitude, request.latitude,
        )
    }

    @ResponseStatus(HttpStatus.OK)
    @GetMapping("/history")
    override fun history(
        @RequestAttribute("memberId") memberId: Long,
    ): List<PlaceHistoryResponse> {
        return PlaceHistoryResponse.fromList(placeFacade.getHistory(memberId))
    }

    @ResponseStatus(HttpStatus.NO_CONTENT)
    @DeleteMapping("/history")
    override fun deleteHistory(
        @RequestAttribute("memberId") memberId: Long,
        @Valid @RequestBody request: PlaceHistoryDeleteRequest,
    ) {
        placeFacade.deleteHistory(memberId, request.searchedAt)
    }
}
